sdk.go

```diff
@@ -151,11 +154,64 @@ func Serve(fn v1beta1.FunctionRunnerServiceServer, o ...ServeOption) error {
 
 	srv := grpc.NewServer(grpc.MaxRecvMsgSize(so.MaxRecvMsgSize), grpc.Creds(so.Credentials))
 	reflection.Register(srv)
-	v1beta1.RegisterFunctionRunnerServiceServer(srv, fn)
+	v1.RegisterFunctionRunnerServiceServer(srv, fn)
+	v1beta1.RegisterFunctionRunnerServiceServer(srv, ServeBeta(fn))
 	return errors.Wrap(srv.Serve(lis), "cannot serve mTLS gRPC connections")
 }
 
 // NewLogger returns a new logger.
 func NewLogger(debug bool) (logging.Logger, error) {
 	return logging.NewLogger(debug)
 }
+
+// A BetaServer is a v1beta1 FunctionRunnerServiceServer that wraps an identical
+// v1 FunctionRunnerServiceServer. This requires the v1 and v1beta1 protos to be
+// identical.
+//
+// Functions were promoted from v1beta1 to v1 in Crossplane v1.17. Crossplane
+// v1.16 and earlier only sends v1beta1 RunFunctionRequests. Functions should
+// use the BetaServer for backward compatibility, to support Crossplane v1.16
+// and earlier.
+type BetaServer struct {
+	wrapped v1.FunctionRunnerServiceServer
+
+	v1beta1.UnimplementedFunctionRunnerServiceServer
+}
+
+// ServeBeta returns a v1beta1.FunctionRunnerServiceServer that wraps the
+// suppled v1.FunctionRunnerServiceServer.
+func ServeBeta(s v1.FunctionRunnerServiceServer) *BetaServer {
+	return &BetaServer{wrapped: s}
+}
+
+// RunFunction calls the RunFunction method of the wrapped
+// v1.FunctionRunnerServiceServer. It converts from v1beta1 to v1 and back by
+// round-tripping through protobuf marshaling.
+func (s *BetaServer) RunFunction(ctx context.Context, req *v1beta1.RunFunctionRequest) (*v1beta1.RunFunctionResponse, error) {
+	gareq := &v1.RunFunctionRequest{}
+
+	b, err := proto.Marshal(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot marshal v1beta1 RunFunctionRequest to protobuf bytes")
+	}
+
+	if err := proto.Unmarshal(b, gareq); err != nil {
+		return nil, errors.Wrap(err, "cannot unmarshal v1 RunFunctionRequest from v1beta1 protobuf bytes")
+	}
+
+	garsp, err := s.wrapped.RunFunction(ctx, gareq)
+	if err != nil {
+		// This error is intentionally not wrapped. This middleware is just
+		// calling an underlying RunFunction.
+		return nil, err
+	}
+
+	rsp := &v1beta1.RunFunctionResponse{}
+	b, err = proto.Marshal(rsp)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot marshal v1beta1 RunFunctionResponse to protobuf bytes")
+	}
+
+	err = proto.Unmarshal(b, garsp)
+	return rsp, errors.Wrap(err, "cannot unmarshal v1 RunFunctionResponse from v1beta1 protobuf bytes")
```

Comment:
i like the wrapping approach we're taking here to seamlessly serve both beta and v1 at the same time.  But I could use one extra check on this logic because I'm not sure if I've got it right just yet...

On line 210 we marshal an empty beta response to bytes.  Then marshal those bytes from the empty beta response into the GA response we got from running the function (does that potentially overwrite fields/values in the response?).

And then we finish by returning the empty beta response that I don't expect to have any content in it. That doesn't seem correct to me, can you help me understand if I'm missing something? 🤔

Reply:
No you're right, that's definitely the wrong way around... which makes me really confused about why it appeared to work when I tested it manually. 🤔